.ci/utils/restore-build-cache.yml

```diff
@@ -1,101 +1,45 @@
-        }
-        downloadArtifactAndContinue() {
-          if [ -z "$LATEST_ART_URL" ]
-          then
-            echo "No latest artifact for merge-target branch found at URL $REST_ART"
-          else
-            curl "$LATEST_ART_URL" > "${BUILD_STAGINGDIRECTORY}/$ART_NAME.zip"
-            PROJECT_DIR=$PWD
-            cd $BUILD_STAGINGDIRECTORY
-            unzip "$ART_NAME.zip"
-            echo "Using Dependency cache for buildID: $LATEST_BUILD_ID"
-            echo "Build log for build that produced the cache: $LATEST_BUILD_PAGE"
-            echo "Build badge for build that produced the cache: $LATEST_BUILD_BADGE"
-            echo "Build artifact from build that produced the cache: $LATEST_ART_URL"
-            echo "Restoring build cache into:"
-            mkdir -p $ESY__CACHE_INSTALL_PATH
-            echo $ESY__CACHE_INSTALL_PATH
-            echo "##vso[task.setvariable variable=esy_export_dir_to_import]${BUILD_STAGINGDIRECTORY}/${ART_NAME}/_export"
-            if [[ -d ${ART_NAME}/_export ]]; then
-              echo "Cached builds to import from ${ART_NAME}/_export in next CI step"
-              ls ${ART_NAME}/_export
-              mv ${ART_NAME}/_export ${PROJECT_DIR}/_export
-            else
-              echo "No _export directory to import from build cache"
-              echo "Here's the contents of build cache:"
-              find ${BUILD_STAGINGDIRECTORY}
-            fi
-          fi
-        }
-        fetchArtifactURL
-        downloadArtifactAndContinue
+  - bash: |
+      # COMPUTE THE ESY INSTALL CACHE LOCATION AHEAD OF TIME
+      DESIRED_LEN="86"
+      # Note: This will need to change when upgrading esy version
+      # that reenables long paths on windows.
+      if [ "$AGENT_OS" == "Windows_NT" ]; then
+        DESIRED_LEN="33"
+      fi
+      HOME_ESY3="$HOME/.esy/3"
+      HOME_ESY3_LEN=${#HOME_ESY3}
+      NUM_UNDERS=$(echo "$(($DESIRED_LEN-$HOME_ESY3_LEN))")
+      UNDERS=$(printf "%-${NUM_UNDERS}s" "_")
+      UNDERS="${UNDERS// /_}"
+      THE_ESY__CACHE_INSTALL_PATH=${HOME_ESY3}${UNDERS}/i
+      if [ "$AGENT_OS" == "Windows_NT" ]; then
+        THE_ESY__CACHE_INSTALL_PATH=$( cygpath --mixed --absolute "$THE_ESY__CACHE_INSTALL_PATH")
+      fi
+      echo "THE_ESY__CACHE_INSTALL_PATH: $THE_ESY__CACHE_INSTALL_PATH"
+      # This will be exposed as an env var ESY__CACHE_INSTALL_PATH, or an
+      # Azure var esy__cache_install_path
+      echo "##vso[task.setvariable variable=esy__cache_install_path]$THE_ESY__CACHE_INSTALL_PATH"
+    displayName: '[Cache] calculate esy store path'
       
-  - powershell: esy.cmd import-dependencies
-    continueOnError: true
-    condition: and(eq(variables['AGENT.OS'], 'Windows_NT'), and(eq(variables['Build.Reason'], 'PullRequest'), and(succeeded(), ne(variables['Build.SourceBranch'], variables['System.PullRequest.TargetBranch']))))
```

Comment:
This stuff was important (esy import-dependencies). Is it moved somewhere else?

Reply:
What happened was Azure changed the location of where our cache directory was on the physical file system, and so artifacts broke in future builds because the artifacts had hard coded paths in them. So I made sure that we actually ran `esy export-dependencies` and `esy import-dependencies` which fixes that issue.

(We can also use the REST API locally from laptops to pull down those caches and warm up our local development).

Reply:
@jordwalke Hm... I didn't realize this was added because of that. The path issue does not seem to be happening now (e.g. https://dev.azure.com/esy-ocaml/esy-ocaml/_build/results?buildId=421). Do you remember if this issue happened in specific situations or how could we repro it?

Also, did it happen in all platforms? I was wondering if maybe the recent upgrade you did to Windows_2019 vm would have fixed that problem.

> We can also use the REST API locally from laptops to pull down those caches and warm up our local development

This is definitely a regression. I wonder if there is a way to expose the cache in the release artifacts? Reintroducing the REST API based flow would mean we would have to own again the issues with permissions, source and target branch handling etc which was one of the nice upsides of using Azure cache.

Reply:
>  I didn't realize this was added because of that. The path issue does not seem to be happening now (e.g. https://dev.azure.com/esy-ocaml/esy-ocaml/_build/results?buildId=421). Do you remember if this issue happened in specific situations or how could we repro it?

Some of the paths unexpectedly changed (the user directory where we have the esy cache for example). I don't believe they make any promises that it will not change unexpectedly. So I put the esy import/export in there because it guards against it ever changing in the future - as well as enabling the ability for us to use the cache to pull down artifacts from CI to warm up our laptops (it can't be done without first esy export-dependencies)

Reply:
> > We can also use the REST API locally from laptops to pull down those caches and warm up our local development

> This is definitely a regression. I wonder if there is a way to expose the cache in the release artifacts? Reintroducing the REST API based flow would mean we would have to own again the issues with permissions, source and target branch handling etc which was one of the nice upsides of using Azure cache.

I think I might not have been clear. What I was saying was that if the cache is in esy export-dependencies form, then it is trivial for local development machines (laptops) to have simple bash scripts that use the REST API and pull them down and magically speed up their local builds. (This is most important for Windows).
I wasn't suggesting that the only way to get the caches in esy export-dependencies form is to use the REST API. It's just that the .ci configurations have diverged and since the new Azure cache started being prototyped outside of this repo, I had already improved some of the steps in this repo's CI config, and then they got out of sync in terms of their features. So I'm asking if this PR can be updated to include the features that I added in this repo, before we merge?

Using `esy import-dependencies` and `esy export-dependencies` were the two that I could think of (I also added a couple of custom steps for windows which delete the package `ocamlfind` from the imported build cache (which we need until my PR for esy lands)).

I had to configure the `esy import-dependencies` a specific way (note my previous config which calls `esy.cmd`). It wouldn't work otherwise.